src/plonk_proof.rs

```diff
@@ -160,63 +160,65 @@ impl<C: HaloCurve> OldProof<C> {
 
 #[derive(Debug, Clone)]
 /// The `Target` version of `OldProof`.
-pub struct OldProofTarget {
-    pub halo_g: AffinePointTarget,
-    pub halo_us: Vec<Target>,
+pub struct OldProofTarget<C: Curve> {
+    pub halo_g: AffinePointTarget<C>,
+    pub halo_us: Vec<Target<C::ScalarField>>,
 }
 
-impl OldProofTarget {
-    pub fn populate_witness<C: HaloCurve>(
+impl<C: HaloCurve> OldProofTarget<C> {
+    pub fn populate_witness(
         &self,
         witness: &mut PartialWitness<C::BaseField>,
         values: &OldProof<C>,
     ) -> Result<()> {
         witness.set_point_target(self.halo_g, values.halo_g);
         debug_assert_eq!(self.halo_us.len(), values.halo_us.len());
         witness.set_targets(
-            &self.halo_us,
+            &Target::convert_slice(&self.halo_us),
             &C::ScalarField::try_convert_all(&values.halo_us)?,
         );
 
         Ok(())
     }
 }
 
-pub struct ProofTarget {
+pub struct ProofTarget<C: Curve, InnerC: Curve<BaseField = C::ScalarField>> {
     /// A commitment to each wire polynomial.
-    pub c_wires: Vec<AffinePointTarget>,
+    pub c_wires: Vec<AffinePointTarget<InnerC>>,
     /// A commitment to Z, in the context of the permutation argument.
-    pub c_plonk_z: AffinePointTarget,
+    pub c_plonk_z: AffinePointTarget<InnerC>,
     /// A commitment to the quotient polynomial.
-    pub c_plonk_t: Vec<AffinePointTarget>,
+    pub c_plonk_t: Vec<AffinePointTarget<InnerC>>,
 
     /// The opening of each polynomial at each `PublicInputGate` index.
-    pub o_public_inputs: Option<Vec<OpeningSetTarget>>,
+    pub o_public_inputs: Option<Vec<OpeningSetTarget<C>>>,
     /// The opening of each polynomial at `zeta`.
-    pub o_local: OpeningSetTarget,
+    pub o_local: OpeningSetTarget<C>,
     /// The opening of each polynomial at `g * zeta`.
-    pub o_right: OpeningSetTarget,
+    pub o_right: OpeningSetTarget<C>,
     /// The opening of each polynomial at `g^65 * zeta`.
-    pub o_below: OpeningSetTarget,
+    pub o_below: OpeningSetTarget<C>,
 
     /// L_i in the Halo reduction.
-    pub halo_l_i: Vec<AffinePointTarget>,
+    pub halo_l_i: Vec<AffinePointTarget<InnerC>>,
     /// R_i in the Halo reduction.
-    pub halo_r_i: Vec<AffinePointTarget>,
+    pub halo_r_i: Vec<AffinePointTarget<InnerC>>,
     /// The purported value of G, i.e. <s, G>, in the context of Halo.
-    pub halo_g: AffinePointTarget,
+    pub halo_g: AffinePointTarget<InnerC>,
     /// The data used in the final Schnorr protocol of the Halo opening proof.
-    pub schnorr_proof: SchnorrProofTarget,
+    pub schnorr_proof: SchnorrProofTarget<InnerC>,
 }
 
-impl ProofTarget {
+impl<C: HaloCurve, InnerC: HaloCurve<BaseField = C::ScalarField, ScalarField = C::BaseField>>
```

Comment:
Does it work OK without `ScalarField = C::BaseField`? I think it'd be best not to require that, in case someone wanted to do bounded recursion e.g. with a Cocks-Pinch curve.

Reply:
(Also comes up a couple times below in `plonk_recursion.rs`)

Reply:
Good point, I made the changes.